core/pkg/monitor/dcgm_exporter.go

```diff
@@ -0,0 +1,385 @@
+	// URL of the Prometheus API endpoint.
+	//
+	// The user must provide the URL of the Prometheus API endpoint and the Prometheus query
+	// to fetch the relevant metrics, for example:
+	// http://prometheus-gateway:9400/api/v1/query?query=DCGM_FI_DEV_GPU_TEMP{node="l1337", cluster="globular"}
+	// It will be parsed into the base URL and the query.
+	URL string
+
+	// HTTP headers to include in the requests.
+	//
+	// Typically used for authentication.
+	Headers map[string]string
+
+	// Logger to use for logging.
+	Logger *observability.CoreLogger
+
+	// Client is the base retryable HTTP client to use for the Prometheus API client.
+	Client *retryablehttp.Client
+}
+
+func NewDCGMExporter(params DCGMExporterParams) *DCGMExporter {
+	// Set up a default retryable client if none is provided.
+	if params.Client == nil {
+		params.Client = retryablehttp.NewClient()
+		params.Client.Logger = params.Logger
+		params.Client.CheckRetry = retryablehttp.ErrorPropagatedRetryPolicy
+		params.Client.RetryMax = DefaultOpenMetricsRetryMax
+		params.Client.RetryWaitMin = DefaultOpenMetricsRetryWaitMin
+		params.Client.RetryWaitMax = DefaultOpenMetricsRetryWaitMax
+		params.Client.HTTPClient.Timeout = DefaultOpenMetricsTimeout
+		params.Client.Backoff = clients.ExponentialBackoffWithJitter
+	}
+
+	// Case 1: /metrics endpoint.
+	if strings.HasSuffix(params.URL, "/metrics") {
+		// TODO: Implement
+		return nil
+	}
+
+	// Case 2: Prometheus API (/api/v1/query) query.
+	var err error
+
+	// Split the URL into the base URL and the query.
+	baseUrl, queries, err := ParsePromQLURL(params.URL)
+	if err != nil {
+		params.Logger.Error("monitor: openmetrics: error parsing URL", "error", err)
+		return nil
+	}
+	params.Logger.Debug("monitor: openmetrics: parsed URL", "url", baseUrl, "queries", queries)
+
+	// Create headers config.
+	headersConfig := &config.Headers{
+		Headers: make(map[string]config.Header),
+	}
+	for key, value := range params.Headers {
+		headersConfig.Headers[key] = config.Header{
+			Values: []string{value},
+		}
+	}
+
+	// Create a chain of RoundTrippers: headers -> retryable client.
+	roundTripper := config.NewHeadersRoundTripper(
+		headersConfig,
+		params.Client.StandardClient().Transport,
+	)
+
+	apiClient, err := api.NewClient(api.Config{
+		Address:      baseUrl,
+		RoundTripper: roundTripper,
+	})
+	if err != nil {
+		params.Logger.Error("monitor: dcgm_exporter: error creating Prometheus API client", "error", err)
+		return nil
+	}
+
+	// Create the Prometheus v1 API client.
+	prometheusAPI := v1.NewAPI(apiClient)
+
+	return &DCGMExporter{
+		baseUrl:       baseUrl,
+		queries:       queries,
+		logger:        params.Logger,
+		prometheusAPI: prometheusAPI,
+	}
+}
+
+// ParsePromQLURL parses a Prometheus API URL to get the base URL and query parameters.
+//
+// parsedURL.Path is omitted as Prometheus' api.Client() assumes /api/v1/query.
+func ParsePromQLURL(fullURL string) (baseURL string, queries []string, err error) {
```

Comment:
`PromQL`?

Reply:
Prometheus Query Language, that's the expected format. Added an explanation.